File: 3713-longest-balanced-substring-i/3713-longest-balanced-substring-i.py
class Solution:
    def longestBalanced(self, s: str) -> int:
        ans = 0
        for i in range(0, len(s)):
            freq = {}
            for j in range(i, len(s)):
                freq[s[j]] = freq.get(s[j], 0) + 1

                balance = True
                curr_freq = freq[s[j]]
                for _, v in freq.items():
                    if curr_freq != v:
                        balance = False
                        break
                
                if balance:
                    ans = max(ans, j - i + 1)
            
        return ans
    
# hash map, string
# time: O(n^2)
# space: O(1)

File: 3713-longest-balanced-substring-i/3713-longest-balanced-substring-i.go
func longestBalanced(s string) int {
	ans := 0
	for i := 0; i < len(s); i++ {
		set := map[byte]int{}
		for j := i; j < len(s); j++ {
			set[s[j]]++

			balance := true
			currFreq := set[s[j]]
			for _, v := range set {
				if currFreq != v {
					balance = false
					break
				}
			}

			if balance {
				ans = max(ans, j-i+1)
			}
		}
	}

	return ans
}

// hash map, string
// time: O(n^2)
// space: O(1)